File: clients/python/chex.py
"""CHEX client — drives the `chex` binary's persistent NDJSON loop.

No pip dependencies. Requires the `chex` binary on PATH or an explicit path.
One long-lived subprocess.

    from chex import CHEX

    with CHEX() as c:
        # schema is a name (with schema_dir) or a path to a .schema.json file
        data = c.validate("./schemas/person.schema.json", {"name": "Ada"})
        # or, resolving "person" against a directory:
        data = c.validate("person", {"name": "Ada"}, schema_dir="./schemas")

`validate` returns the validated data on success and raises CHEXError when the
data does not match the schema. `request(op)` is a raw escape hatch returning
the full response dict.
"""

import json
import subprocess
import threading


class CHEXError(RuntimeError):
    pass


class CHEX:
    def __init__(self, binary="chex"):
        self._proc = subprocess.Popen(
            [binary, "exec", "--loop"],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            text=True,
            bufsize=1,
        )
        self._lock = threading.Lock()

    def request(self, op):
        """Send one raw machine-protocol op; return the full response dict."""
        line = json.dumps(op)
        with self._lock:  # ponytail: one call in flight; drop the lock only if you pipeline
            if self._proc.poll() is not None:
                raise CHEXError("chex process has exited")
            self._proc.stdin.write(line + "\n")
            self._proc.stdin.flush()
            reply = self._proc.stdout.readline()
        if not reply:
            raise CHEXError("chex closed the stream (stderr may have details)")
        return json.loads(reply)

    def _op(self, op, **fields):
        payload = {"op": op}
        for key, value in fields.items():
            if value is not None:
                payload[key] = value
        response = self.request(payload)
        if not response.get("ok"):
            raise CHEXError((response.get("error") or {}).get("message", "chex error"))
        return response.get("result")

    def validate(self, schema, data, schema_dir=None):
        """Validate data against a schema (name or .schema.json path).

        Returns the validated data; raises CHEXError if it does not match.
        """
        return self._op("validate", schema=schema, data=data, schemaDir=schema_dir)

    def close(self):
        if self._proc.poll() is None:
            self._proc.stdin.close()
            self._proc.wait()

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

File: clients/swift/Chex.swift
// CHEX client — drives the `chex` binary's persistent NDJSON loop.
//
// Foundation only, no SwiftPM dependencies. Requires the `chex` binary on PATH
// or an explicit path. One long-lived subprocess.
//
//   let c = try CHEX()
//   defer { c.close() }
//   let data = try c.validate("./schemas/person.schema.json", ["name": "Jane Doe", "age": 30])
//   // name form, resolved against a directory:
//   try c.validate("person", ["name": "Jane Doe", "age": 30], schemaDir: "./schemas")
//
// `validate` returns the validated data and throws CHEXError on a schema
// mismatch. Method names follow Swift's camelCase. `request(_:)` is a raw escape
// hatch returning the full response dictionary.

import Foundation

struct CHEXError: Error { let message: String }

final class CHEX {
    private let proc = Process()
    private let stdinPipe = Pipe()
    private let stdoutPipe = Pipe()
    private let lock = NSLock()
    private var buffer = Data()

    init(binary: String = "chex") throws {
        // /usr/bin/env resolves `binary` against PATH.
        proc.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        proc.arguments = [binary, "exec", "--loop"]
        proc.standardInput = stdinPipe
        proc.standardOutput = stdoutPipe
        try proc.run()
    }

    /// Send one raw machine-protocol op; return the full response dictionary.
    func request(_ op: [String: Any]) throws -> [String: Any] {
        let payload = try JSONSerialization.data(withJSONObject: op)
        lock.lock() // ponytail: one call in flight; drop the lock only if you pipeline
        defer { lock.unlock() }
        guard proc.isRunning else { throw CHEXError(message: "chex process has exited") }
        let handle = stdinPipe.fileHandleForWriting
        handle.write(payload)
        handle.write(Data([0x0a]))
        let line = try readLine()
        guard let object = try JSONSerialization.jsonObject(with: Data(line.utf8)) as? [String: Any] else {
            throw CHEXError(message: "chex returned a non-object response")
        }
        return object
    }

    /// Validate `data` against a schema (name or .schema.json path). Returns the validated data.
    @discardableResult
    func validate(_ schema: String, _ data: [String: Any], schemaDir: String? = nil) throws -> Any {
        var op: [String: Any] = ["op": "validate", "schema": schema, "data": data]
        if let schemaDir = schemaDir { op["schemaDir"] = schemaDir }
        let response = try request(op)
        guard (response["ok"] as? Bool) == true else {
            let message = (response["error"] as? [String: Any])?["message"] as? String ?? "chex error"
            throw CHEXError(message: message)
        }
        return response["result"] as Any
    }

    /// Close stdin so the loop ends, and wait for the process to exit.
    func close() {
        guard proc.isRunning else { return }
        stdinPipe.fileHandleForWriting.closeFile()
        proc.waitUntilExit()
    }

    // Read one newline-delimited line from stdout, buffering across reads.
    private func readLine() throws -> String {
        while true {
            if let index = buffer.firstIndex(of: 0x0a) {
                let lineData = buffer[buffer.startIndex..<index]
                buffer.removeSubrange(buffer.startIndex...index)
                return String(decoding: lineData, as: UTF8.self)
            }
            let chunk = stdoutPipe.fileHandleForReading.availableData
            if chunk.isEmpty { throw CHEXError(message: "chex closed the stream (stderr may have details)") }
            buffer.append(chunk)
        }
    }
}
